### PR TITLE
ChainSync client test: model pipelining behavior

### PR DESCRIPTION
**Pull request by amesgen**
Closes [CAD-4192](https://input-output.atlassian.net/browse/CAD-4192)

### CODE REVIEW

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs [hunk 1]
```diff
@@ -435,6 +466,14 @@ runChainSync securityParam (ClientUpdates clientUpdates)
       tick <- atomically $ LogicalClock.getCurrentTick clock
       traceWith tr (tick, ev)
 
+    catchChainSyncClient :: m a -> (ChainSyncClientException -> m a) -> m a
+    catchChainSyncClient ma handler = ma `catch` \case
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:470]: Minor: maybe `catches` would be more idiomatic/recognizable? It'd save one layer of `fromException`, but not both layers.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs [hunk 2]
```diff
@@ -201,11 +206,19 @@ toChainUpdates = concatMap $ \case
...
-  ClientUpdates { getClientUpdates :: Schedule [ChainUpdate] }
+  ClientUpdates { getClientUpdates :: Schedule ChainUpdate }
   deriving (Show)
 
 newtype ServerUpdates =
-  ServerUpdates { getServerUpdates :: Schedule [ChainUpdate] }
+  ServerUpdates { getServerUpdates :: Schedule ChainUpdate }
+  deriving (Show)
+
+-- | A 'Schedule' of when events when we learn that a specific block is invalid.
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:216]: typo: "when events when we learn"

Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:216]: typo: "when events when we learn"

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs [hunk 3]
```diff
@@ -510,16 +541,18 @@ data ChainSyncClientSetup = ChainSyncClientSetup
...
 instance Arbitrary ChainSyncClientSetup where
   arbitrary = do
     securityParam  <- SecurityParam <$> choose (2, 5)
     clientUpdates0 <- evalStateT
-      (ClientUpdates <$> genUpdateSchedule securityParam)
+      (ClientUpdates <$> genUpdateSchedule SelectedChainBehavior securityParam)
       emptyUpdateState
-    serverUpdates  <- evalStateT
-      (ServerUpdates <$> genUpdateSchedule securityParam)
+    serverUpdates <- evalStateT
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:554]: Minor: preserve the aligning space?

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs [hunk 4]
```diff
@@ -570,8 +613,11 @@ instance Show ChainSyncClientSetup where
       , "serverUpdates:"
       , ppSchedule ppChainUpdate (getServerUpdates serverUpdates) <> "--"
       , "startTick: " <> show startTick
+      , "invalidBlocks: "
+      , ppSchedule condense (getInvalidBlocks invalidBlocks)
       ]
 
+
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:620]: Minor: extra newline

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs [hunk 5]
```diff
@@ -669,15 +716,21 @@ getChainUpdates = reverse . cusUpdates
 -- '_currentChain'.
 prop_genChainUpdates :: SecurityParam -> Int -> Property
 prop_genChainUpdates securityParam n =
-    forAll (genChainUpdates securityParam n emptyUpdateState) $ \cus ->
+    forAll (genChainUpdates SelectedChainBehavior securityParam n emptyUpdateState) $ \cus ->
       Chain.applyChainUpdates (toChainUpdates (getChainUpdates cus)) Genesis ===
       Just (cusCurrentChain cus)
 
-genChainUpdates :: SecurityParam
+data UpdateBehavior =
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:723]: I anticipate someone later will benefit from a bit of Haddock on this type/these constructors 😄

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs [hunk 6]
```diff
@@ -688,38 +741,58 @@ genChainUpdates securityParam n =
...
-        Nothing      -> firstBlock <$> genForkNo
-        Just curHead -> do
-          forkNo <- genForkNo
-          return $ modifyFork (const forkNo) (successorBlock curHead)
-      modify $ setChain (Chain.addBlock block chain)
-      return block
-
-    genAddBlock = do
-      block <- genBlockToAdd
+    genAddBlock isValid = do
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:780]: I liked `validity` (as above) more than `isValid` for this parameter.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs [hunk 7]
```diff
@@ -688,38 +741,58 @@ genChainUpdates securityParam n =
...
 
-    genSwitchFork  = do
+    genSwitchFork genRollBackBlocks = do
       ChainUpdateState { cusCurrentChain = chain } <- get
-      rollBackBlocks <- lift $ choose (1, k)
+      rollBackBlocks <- lift genRollBackBlocks
       let chain' = Chain.drop rollBackBlocks chain
       modify $ setChain chain'
-      blocks <- replicateM rollBackBlocks genBlockToAdd
+      blocks <- replicateM rollBackBlocks (genBlockToAdd Valid)
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:789]: Do we _never_ want to switch from one invalid fork to another invalid fork?

I imagine we discussed this on our phone call last week, but I do not remember (got spiked 😞). So, if we never do: we should comment here explaining why :)

**Review by nfrisby — CHANGES_REQUESTED** [review]
This looks very nice. Thanks for the very well-organized commits!

Requesting Changes for a few more Haddock comments, mainly. I also ask one relatively-open-ended question about the design.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs (hunk 7 above)
Comment (amesgen) [reply ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:789]: I don't think we discussed this, but it it indeed possible honest behavior to switch to a *better* invalid fork (i.e. we receive a better trap header). I will add this possibility later today.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs (hunk 1 above)
Comment (amesgen) [reply ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs]: Agreed, I took the opportunity to slightly simplify and generalize this function :+1:

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs (hunk 7 above)
Comment (nfrisby) [reply ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs:800]: We discussed this on a call and decided to include this expressivity boost in https://input-output.atlassian.net/browse/CAD-4314

So I'll resolve this comment thread.

**Review by nfrisby — APPROVED** [review]
All set. Approved! Thanks.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/ChainSync/Client.hs]: Ah, my GitHub UI seems bugged on this comment -- please ignore. It was addressed.

**Comment by amesgen** [comment]
bors merge

**Comment by amesgen** [comment]
bors retry

**Pull request merged**